rflx/pyrflx/message.py

```diff
@@ -0,0 +1,234 @@
+        if fld == model.FINAL.name:
+            return ""
+        for l in self._model.outgoing(model.Field(fld)):
+            if self.__simplified(l.condition) == TRUE:
+                return l.target.name
+        return ""
+
+    def _prev_field(self, fld: str) -> str:
+        if fld == model.INITIAL.name:
+            return ""
+        for l in self._model.incoming(model.Field(fld)):
+            if self.__simplified(l.condition) == TRUE:
+                return l.source.name
+        return ""
+
+    def _get_length(self, fld: str) -> Number:
+        prv = self._prev_field(fld)
+        for l in self._model.outgoing(model.Field(prv)):
+            length = self.__simplified(l.length)
+            if isinstance(length, Number):
+                return length
+        typeval = self._fields[fld].typeval
+        if isinstance(typeval, ScalarValue):
+            return Number(typeval.size)
+        raise RuntimeError(f"cannot determine length of {fld}")
+
+    def _has_length(self, fld: str) -> bool:
+        prv = self._prev_field(fld)
+        for l in self._model.outgoing(model.Field(prv)):
+            if isinstance(self.__simplified(l.length), Number):
+                return True
+        return isinstance(self._fields[fld].typeval, ScalarValue)
+
+    def _get_first(self, fld: str) -> Number:
+        for l in self._model.incoming(model.Field(fld)):
+            if self.__simplified(l.condition) == TRUE and l.first != UNDEFINED:
+                first = self.__simplified(l.first)
+                break
+        else:
+            prv = self._prev_field(fld)
+            first = self.__simplified(Add(self._fields[prv].first, self._fields[prv].length))
+        if isinstance(first, Number):
+            return first
+        raise RuntimeError(f"cannot determine first of {fld}")
+
+    def _has_first(self, fld: str) -> bool:
+        prv = self._prev_field(fld)
+        first = self.__simplified(Add(self._fields[prv].first, self._fields[prv].length))
+        return isinstance(first, Number)
+
+    def set(self, fld: str, value: Any) -> None:
+        if not isinstance(value, self._fields[fld].typeval.accepted_type):
+            raise TypeError(
+                f"cannot assign different types: {self._fields[fld].typeval.accepted_type.__name__}"
+                f" != {type(value).__name__}"
+            )
+        try:
+            field = self._fields[fld]
+            field.first = self._get_first(fld)
+            field.length = self._get_length(fld)
+            self._fields[fld].typeval.assign(value, True)
+            if all(
+                [
+                    self.__simplified(o.condition) == FALSE
+                    for o in self._model.outgoing(model.Field(fld))
+                ]
+            ):
+                self._fields[fld].typeval.clear()
+                raise ValueError("value does not allow path to next field")
+        except KeyError:
+            raise KeyError(f"cannot access field {fld}")
```

Comment:
I think catching exceptions should be avoided. Here it should be easy to check if `fld in self._fields`.